File: javascript/cfg.js
function firstSetRaHesabKon(grammar) {
    let firstSet = {};
    let epsilon = 'Є';

    for (let nonTerminal in grammar) {
        firstSet[nonTerminal] = new Set();
    }

    let hasChanged = true;
    while (hasChanged) {
        hasChanged = false;
        for (let nonTerminal in grammar) {
            let productions = grammar[nonTerminal];
            let oldSize = firstSet[nonTerminal].size;
            for (let production of productions) {
                let symbols = production.split('');
                let allSymbolsCanProduceEpsilon = true;
                for (let symbol of symbols) {
                    if (grammar[symbol]) {
                        for (let s of firstSet[symbol]) {
                            if (s !== epsilon) {
                                firstSet[nonTerminal].add(s);
                            }
                        }
                        if (!firstSet[symbol].has(epsilon)) {
                            allSymbolsCanProduceEpsilon = false;
                            break;
                        }
                    } else {
                        firstSet[nonTerminal].add(symbol);
                        allSymbolsCanProduceEpsilon = false;
                        break;
                    }
                }
                if (allSymbolsCanProduceEpsilon) {
                    firstSet[nonTerminal].add(epsilon);
                }
            }
            if (firstSet[nonTerminal].size !== oldSize) {
                hasChanged = true;
            }
        }
    }

    return firstSet;
}

let grammar1 = {
    'S': ['aBDh'],
    'B': ['cC'],
    'C': ['bC', 'Є'],
    'D': ['EF'],
    'E': ['g', 'Є'],
    'F': ['f', 'Є']
};

let grammar2 = {
    'S': ['ACB', 'Cbb', 'Ba'],
    'A': ['da', 'BC'],
    'B': ['g', 'Є'],
    'C': ['h', 'Є']
};

let grammar3 = {
    'E': ['TE’'],
    'E’': ['+TE’', 'Є'],
    'T': ['FT’'],
    'T’': ['*FT’', 'Є'],
    'F': ['(E)', 'i']
};

function followSetRaHesabKon(grammar, firstSet) {
    let followSet = {};
    let epsilon = 'Є';
    let startSymbol = Object.keys(grammar)[0];

    for (let nonTerminal in grammar) {
        followSet[nonTerminal] = new Set();
    }

    followSet[startSymbol].add('$');

    let hasChanged = true;
    while (hasChanged) {
        hasChanged = false;
        for (let nonTerminal in grammar) {
            let productions = grammar[nonTerminal];
            for (let production of productions) {
                let symbols = production.split('');
                for (let i = 0; i < symbols.length; i++) {
                    if (grammar[symbols[i]]) { 
                        let oldSize = followSet[symbols[i]].size;
                        if (i < symbols.length - 1) { 
                            if (grammar[symbols[i + 1]]) { 
                                for (let s of firstSet[symbols[i + 1]]) {
                                    if (s !== epsilon) {
                                        followSet[symbols[i]].add(s);
                                    }
                                }
                                if (firstSet[symbols[i + 1]].has(epsilon)) {
                                    for (let j = i + 2; j < symbols.length; j++) {
                                        if (grammar[symbols[j]]) {
                                            for (let s of firstSet[symbols[j]]) {
                                                if (s !== epsilon) {
                                                    followSet[symbols[i]].add(s);
                                                }
                                            }
                                            if (!firstSet[symbols[j]].has(epsilon)) {
                                                break;
                                            }
                                        } else {
                                            followSet[symbols[i]].add(symbols[j]);
                                            break;
                                        }
                                    }
                                }
                            } else {
                                followSet[symbols[i]].add(symbols[i + 1]);
                            }
                        }
                        if (i === symbols.length - 1 || symbols.slice(i + 1).every(symbol => firstSet[symbol]?.has(epsilon))) {
                            for (let s of followSet[nonTerminal]) {
                                followSet[symbols[i]].add(s);
                            }
                        }
                        if (followSet[symbols[i]].size !== oldSize) {
                            hasChanged = true;
                        }
                    }
                }
            }
        }
    }

    return followSet;
}

let exampleGrammar = {
    'S': ['AB'],
    'A': ['Sx', 'x', 'Є'],
    'B': ['yB', 'y'],
    'C': ['ACA', 'z']
};

console.log("first: ", firstSetRaHesabKon(exampleGrammar));
console.log("follow: ", followSetRaHesabKon(exampleGrammar, firstSetRaHesabKon(exampleGrammar)));

console.log("grammar1: ", followSetRaHesabKon(grammar1, firstSetRaHesabKon(grammar1)));
console.log("grammar2: ", followSetRaHesabKon(grammar2, firstSetRaHesabKon(grammar2)));
console.log("grammar3: ", followSetRaHesabKon(grammar3, firstSetRaHesabKon(grammar3)));

File: csharp/cfg.cs
using System;
using System.Collections.Generic;
using System.Linq;

public class Program
{
    public static void Main()
    {
        var grammar = new Dictionary<string, string[]>
        {
            {"S", new[] {"ACB", "Cbb", "Ba"}},
            {"A", new[] {"da", "BC"}},
            {"B", new[] {"g", "Є"}},
            {"C", new[] {"h", "Є"}}
        };

        var firstSet = calculateFirstSet(grammar);

        foreach (var pair in firstSet)
        {
            Console.Write("First set of " + pair.Key + ": ");
            foreach (var symbol in pair.Value)
            {
                Console.Write(symbol + " ");
            }
            Console.WriteLine();
        }
		
		var followSet = calculateFollowSet(grammar, firstSet);
		Console.WriteLine("---------------------");
		foreach (var pair in followSet)
        {
            Console.Write("First set of " + pair.Key + ": ");
            foreach (var symbol in pair.Value)
            {
                Console.Write(symbol + " ");
            }
            Console.WriteLine();
        }
    }

    public static Dictionary<string, HashSet<string>> calculateFirstSet(Dictionary<string, string[]> grammar)
    {
        var firstSet = new Dictionary<string, HashSet<string>>();
        const string epsilon = "Є";

        foreach (var nonTerminal in grammar.Keys)
        {
            firstSet[nonTerminal] = new HashSet<string>();
        }

        bool hasChanged;
        do
        {
            hasChanged = false;
            foreach (var nonTerminal in grammar.Keys)
            {
                var productions = grammar[nonTerminal];
                var oldSize = firstSet[nonTerminal].Count;
                foreach (var production in productions)
                {
                    var symbols = production.ToCharArray();
                    var allSymbolsCanProduceEpsilon = true;
                    foreach (var symbol in symbols)
                    {
                        if (grammar.ContainsKey(symbol.ToString()))
                        {
                            foreach (var s in firstSet[symbol.ToString()])
                            {
                                if (s != epsilon)
                                {
                                    firstSet[nonTerminal].Add(s);
                                }
                            }
                            if (!firstSet[symbol.ToString()].Contains(epsilon))
                            {
                                allSymbolsCanProduceEpsilon = false;
                                break;
                            }
                        }
                        else
                        {
                            firstSet[nonTerminal].Add(symbol.ToString());
                            allSymbolsCanProduceEpsilon = false;
                            break;
                        }
                    }
                    if (allSymbolsCanProduceEpsilon)
                    {
                        firstSet[nonTerminal].Add(epsilon);
                    }
                }
                if (firstSet[nonTerminal].Count != oldSize)
                {
                    hasChanged = true;
                }
            }
        } while (hasChanged);

        return firstSet;
    }
	
	public static Dictionary<string, HashSet<string>> calculateFollowSet(Dictionary<string, string[]> grammar, Dictionary<string, HashSet<string>> firstSet)
    {
        var followSet = new Dictionary<string, HashSet<string>>();
        const string epsilon = "Є";
        var startSymbol = grammar.Keys.First();

        foreach (var nonTerminal in grammar.Keys)
        {
            followSet[nonTerminal] = new HashSet<string>();
        }

        followSet[startSymbol].Add("$");

        bool hasChanged;
        do
        {
            hasChanged = false;
            foreach (var nonTerminal in grammar.Keys)
            {
                var productions = grammar[nonTerminal];
                foreach (var production in productions)
                {
                    var symbols = production.ToCharArray();
                    for (int i = 0; i < symbols.Length; i++)
                    {
                        if (grammar.ContainsKey(symbols[i].ToString()))
                        {
                            var oldSize = followSet[symbols[i].ToString()].Count;
                            if (i < symbols.Length - 1)
                            {
                                if (grammar.ContainsKey(symbols[i + 1].ToString()))
                                {
                                    foreach (var s in firstSet[symbols[i + 1].ToString()])
                                    {
                                        if (s != epsilon)
                                        {
                                            followSet[symbols[i].ToString()].Add(s);
                                        }
                                    }
                                    if (firstSet[symbols[i + 1].ToString()].Contains(epsilon))
                                    {
                                        for (int j = i + 2; j < symbols.Length; j++)
                                        {
                                            if (grammar.ContainsKey(symbols[j].ToString()))
                                            {
                                                foreach (var s in firstSet[symbols[j].ToString()])
                                                {
                                                    if (s != epsilon)
                                                    {
                                                        followSet[symbols[i].ToString()].Add(s);
                                                    }
                                                }
                                                if (!firstSet[symbols[j].ToString()].Contains(epsilon))
                                                {
                                                    break;
                                                }
                                            }
                                            else
                                            {
                                                followSet[symbols[i].ToString()].Add(symbols[j].ToString());
                                                break;
                                            }
                                        }
                                    }
                                }
                                else
                                {
                                    followSet[symbols[i].ToString()].Add(symbols[i + 1].ToString());
                                }
                            }
                            if (i == symbols.Length - 1 || symbols.Skip(i + 1).All(symbol => firstSet.ContainsKey(symbol.ToString()) && (firstSet[symbol.ToString()]?.Contains(epsilon) ?? false)))
                            {
                                foreach (var s in followSet[nonTerminal])
                                {
                                    followSet[symbols[i].ToString()].Add(s);
                                }
                            }
                            if (followSet[symbols[i].ToString()].Count != oldSize)
                            {
                                hasChanged = true;
                            }
                        }
                    }
                }
            }
        } while (hasChanged);

        return followSet;
    }

}